### DISCUSSION TITLE
Select Option Groups

### DISCUSSION THREAD

**Discussion by mikebronner** [livewire/flux#354] [Feature Requests]
I couldn't find documentation or code that supported select option groups. Did I miss something? I am using the following work-around:

```blade
                    <flux:select
                        class="!font-sans !w-auto !border-transparent"
                        data-tippy-content="Select A Book"
                        data-tippy-placement="top"
                        wire:model="selectedBook"
                    >

                        @foreach ($bookOptions as $label => $options)
                            <optgroup
                                label="{{ $label }}"
                            >
                                
                                @foreach ($options as $option)
                                    <flux:option
                                        wire:key="{{ $option }}"
                                        value="{{ $option }}"
                                    >
                                        {{ $option }}
                                    </flux:option>
                                @endforeach

                            </optgroup>
                        @endforeach

                    </flux:select>
```

If this is the expected implementation, is this something that would be useful to document? Or is this a feature that could be added?

**Comment by iliyto** [comment]
I think you can use any HTML tags inside `<flux:select>`.  

### Example:

```html
<flux:select
    class="!font-sans !w-auto !border-transparent"
    data-tippy-content="Select A Book"
    data-tippy-placement="top"
    wire:model="selectedBook"
>
    @foreach ($bookOptions as $label => $options)
        <div class=“font-semibold uppercase border-b pb-1” >{{ $label }}</div>
            @foreach ($options as $option)
                <flux:option
                    wire:key="{{ $option }}"
                    value="{{ $option }}"
                >
                    {{ $option }}
                </flux:option>
            @endforeach
        </optgroup>
    @endforeach
</flux:select>

**Comment by seabasss** [comment]
> I couldn't find documentation or code that supported select option groups. Did I miss something? I am using the following work-around:
> 
> ```
>                     <flux:select
>                         class="!font-sans !w-auto !border-transparent"
>                         data-tippy-content="Select A Book"
>                         data-tippy-placement="top"
>                         wire:model="selectedBook"
>                     >
> 
>                         @foreach ($bookOptions as $label => $options)
>                             <optgroup
>                                 label="{{ $label }}"
>                             >
>                                 
>                                 @foreach ($options as $option)
>                                     <flux:option
>                                         wire:key="{{ $option }}"
>                                         value="{{ $option }}"
>                                     >
>                                         {{ $option }}
>                                     </flux:option>
>                                 @endforeach
> 
>                             </optgroup>
>                         @endforeach
> 
>                     </flux:select>
> ```
> 
> If this is the expected implementation, is this something that would be useful to document? Or is this a feature that could be added?

That doesn't work on searchable selects right?

I tried @iliyto's suggestion with div (minus the hanging optgroup) and it shows up, but also stays when you search.

How have other people solved this for searchable selects?

**Comment by ultimateBusiness** [comment]
You can make use of "flux:seperator" inside "flux:select" as a work around to emulate the optgroup behavior.

**Reply by seabasss** [reply]
Do you have an example? Would they not show up if a search doesn’t match?

**Reply by ultimateBusiness** [reply]
<flux:select>
  <flux:separator text="..." />
  <flux:select.option> ...

  <flux:separator text="..." />
  <flux:select.option> ...
</flux:select>

And yes - the separator show's up - but at the moment better than nothing.

A real implementation of optgroup would be better ;-)

**Comment by mauritskorse** [comment]
You could use some css for this:
```blade
<optgroup class="hidden has-[option]:block"> 
   //...
</optgroup>
```

ie:
```blade
@php
        $bookOptions = [
            'Fiction' => ['The Great Gatsby', 'To Kill a Mockingbird'],
            'Non-Fiction' => ['Sapiens', 'Educated'],
            'Science Fiction' => [], 
            'Fantasy' => ['The Hobbit', 'Harry Potter'],
            'Mystery' => [],
            'Romance' => [],
        ];

        $selectedBook = null;
    @endphp

    <flux:select
        class="!font-sans !w-auto !border-transparent"
        wire:model="selectedBook"
    >

        @foreach ($bookOptions as $label => $options)
            <optgroup
                class="hidden has-[option]:block"
                label="{{ $label }}"
            >
                @foreach ($options as $option)
                    <flux:select.option
                        wire:key="{{ $option }}"
                        value="{{ $option }}"
                    >
                        {{ $option }}
                    </flux:select.option>
                @endforeach

            </optgroup>
        @endforeach

    </flux:select>
```

**Reply by seabasss** [reply]
Does not work for variant="listbox" searchable

**Comment by joffreypersia** [comment]
It's hard to believe some basic stuff like `optgroup` on `listbox` are not supported by Flux, but the team does a lot of effort on sidebar. 

A form component should be the priority #1 for mass adoption...

**Comment by helderneves91** [comment]
+1

**Comment by cladico** [comment]
+1

**Comment by cladico** [comment]
I keep coming back to this.
Can we have this feature?
